# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: dom/media/MediaTrackGraph.cpp
# Commit: d6724dbc6875
# Full Hash: d6724dbc68750ecb395016e1cd17151701340f31
# Author: Paul Adenot <paul@paul.cx>
# Date: 2025-01-28 16:18:37
# Description:
#   Bug 1943719 - Repeat bailout logic after dispatching back to the main thread in MediaTrackGraph::DeviceChanged. r=pehrsons
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D235605
# ==============================================================================

diff -r 8cc0b0adac80 -r d6724dbc6875 dom/media/MediaTrackGraph.cpp
--- a/dom/media/MediaTrackGraph.cpp	Tue Jan 28 09:36:30 2025 +0000
+++ b/dom/media/MediaTrackGraph.cpp	Tue Jan 28 09:59:47 2025 +0000
@@ -1033,7 +1033,8 @@
 
   // Dispatch to the bg thread to do the (potentially expensive) query of the
   // maximum channel count, and then dispatch back to the main thread, then to
-  // the graph, with the new info.
+  // the graph, with the new info. The "special case" above is to be handled
+  // back on the main thread as well for the same reasons.
   RefPtr<MediaTrackGraphImpl> self = this;
   NS_DispatchBackgroundTask(NS_NewRunnableFunction(
       "MaxChannelCountUpdateOnBgThread", [self{std::move(self)}]() {
@@ -1055,6 +1056,13 @@
                 MediaTrackGraphImpl* mGraphImpl;
                 uint32_t mMaxChannelCount;
               };
+
+              if (self->mMainThreadTrackCount == 0 &&
+                  self->mMainThreadPortCount == 0) {
+                // See comments above.
+                return;
+              }
+
               self->AppendMessage(
                   MakeUnique<MessageToGraph>(self, maxChannelCount));
             }));
@@ -1620,9 +1628,10 @@
   return false;
 }
 
-auto MediaTrackGraphImpl::OneIteration(
-    GraphTime aStateTime, GraphTime aIterationEnd,
-    MixerCallbackReceiver* aMixerReceiver) -> IterationResult {
+auto MediaTrackGraphImpl::OneIteration(GraphTime aStateTime,
+                                       GraphTime aIterationEnd,
+                                       MixerCallbackReceiver* aMixerReceiver)
+    -> IterationResult {
   if (mGraphRunner) {
     return mGraphRunner->OneIteration(aStateTime, aIterationEnd,
                                       aMixerReceiver);
